clang/lib/Sema/SemaStmtAttr.cpp

```diff
@@ -322,6 +322,71 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
   return ::new (S.Context) UnlikelyAttr(S.Context, A);
 }
 
+CodeAlignAttr *Sema::BuildCodeAlignAttr(const AttributeCommonInfo &CI,
+                                        Expr *E) {
+  if (!E->isValueDependent()) {
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return nullptr;
+    E = Res.get();
+
+    // This attribute requires an integer argument which is a constant power of
+    // two between 1 and 4096 inclusive.
+    if (ArgVal < CodeAlignAttr::MinimumAlignment ||
+	ArgVal > CodeAlignAttr::MaximumAlignment ||
+        (ArgVal < CodeAlignAttr::MaximumAlignment && !ArgVal.isPowerOf2())) {
+      if (std::optional<int64_t> Value = ArgVal.trySExtValue())
+        Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
+            << CI << CodeAlignAttr::MinimumAlignment
+            << CodeAlignAttr::MaximumAlignment << Value.value();
+      else
+        Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
+            << CI << CodeAlignAttr::MinimumAlignment
+            << CodeAlignAttr::MaximumAlignment << E;
+      return nullptr;
+    }
+  }
+  return new (Context) CodeAlignAttr(Context, CI, E);
+}
+
+static Attr *handleCodeAlignAttr(Sema &S, Stmt *St, const ParsedAttr &A) {
+
+  Expr *E = A.getArgAsExpr(0);
+  return S.BuildCodeAlignAttr(A, E);
+}
+
+// Emit duplicate error for [[clang::code_align()]] attribute.
+static void CheckForDuplicateCodeAlignAttrs(Sema &S,
+                                            ArrayRef<const Attr *> Attrs) {
+  // Create a list of CodeAlign attributes only.
+  SmallVector<const CodeAlignAttr *, 8> OnlyCodeAlignAttrs;
```

Comment:
Sorry, just looked at this again.  What is the point of going through all of this?  I would expect a simpler algorithm to be (obviously pseudo code):


```
auto *FirstItr = std::find_if(Attrs.begin(), Attrs.end(), [](const Attr *A) { return isa<const CodeAlignAttr>(A); });
if (FirstItr == Attrs.end()) // no attributes found
  return;
auto *SecondItr = std::find_if(FirstItr + 1, Attrs.end(), [](const Attr *A) { return isa<const CodeAlignAttr>(A); });
if (SecondItr == Attrs.end()) // only 1 attribute found)
  return;

// diagnose here.

```

If you still want to diagnose EVERY duplicate, you can put the 2nd find_if in a loop there instead of as an if.

Reply:
@erichkeane, i have updated duplicate attribute values check with commit https://github.com/llvm/llvm-project/pull/70762/commits/a8b13eb4846d6f1cb41b12088aa371b3fe4edd24 in a simpler way.

>>What is the point of going through all of this?

I could not able figure out in a simple way, so ended up going through all of this while using llvm::find_if()

Reply:
Ah, so the suggestion was something like I just suggested above: Do a `find_if` on the whole collection to find the 'first' attribute.  Then do a `find_if` on the rest of the range (perhaps in a loop) to find a 2nd /further one.  See the code I put above.

Reply:
Thanks @erichkeane. This is fixed with commit https://github.com/llvm/llvm-project/pull/70762/commits/09b00c761f177c3c6ad2285fa5b7642f9c7d28cc

Reply:
Does this new commit https://github.com/llvm/llvm-project/commit/09b00c761f177c3c6ad2285fa5b7642f9c7d28cc
address the concerns ? @erichkeane , @AaronBallman

Reply:
> Because this is a statement attribute, I don't think checking for duplicates is necessary or warrants this much effort. We check for conflicting duplicates on declaration attributes because redeclarations can easily introduce additional attributes. But the same is not true for statements -- the only way to get a duplicate there is with significant effort and it seems like "first/last one wins" would be fine to document. WDYT?

Given how common it is to wrap attributes in macros, and how common it is to accidentally introduce duplicates that way, I still think there is value in diagnosing that a duplicate was ignored.

Reply:
I don't think I've ever seen this happen outside of a test suite. Most attributes hidden behind macros are of this variety:
```
#define FOO __attribute__((foo))
```
so someone would have to do:
```
FOO FOO for(;;);
```
or similar. It's not impossible, certainly. But also not something I'm particularly worried about if it's a slog to diagnose.

Reply:
The ones I see are mostly on declarations (admittedly), but I've definitely seen something akin to:

```
#define FANCY_LOOP __attribute__((foo)) __attribute__((bar)) ANOTHER_MACRO_ATTR
#define MAKE_PARALLEL_LOOP __attribute((baz)) __attribute((bar)) ANOTHER_MACRO_ATTR2

FANCY_LOOP MAKE_PARALLEL_LOOP for(;;);

// Or even...
#define FANCY_PARALLEL_LOOP FANCY_LOOP MAKE_PARALLEL_LOOP

FANCY_PARALLLEL_LOOP
for(;;);
```

Reply:
Also note, if we do NOT diagnose duplicates, we likely have to make sure that we guarantee we pick the first (or last) lexically (likely by not adding them anyway), the `Attrs` collections have changed ordering a few times, so I would hate to try to correct something like that and cause a silent breakage.

Reply:
Yeah, still not something I'm all that worried about, tbh. But I won't stand in the way of good diagnostic behavior!